CMakeLists.txt

```diff
@@ -0,0 +1,225 @@
+if(MOLD_USE_MIMALLOC)
+    if(MOLD_USE_SYSTEM_MIMALLOC)
+        find_library(MIMALLOC_LIBRARIES mimalloc REQUIRED)
+        add_library(mimalloc INTERFACE)
+        target_link_libraries(mimalloc INTERFACE ${MIMALLOC_LIBRARIES})
+        target_compile_definitions(mimalloc INTERFACE USE_SYSTEM_MIMALLOC)
+    else()
+        function(mold_add_mimalloc)
+            string(APPEND CMAKE_C_FLAGS -DMI_USE_ENVIRON=0)
+            option(MI_BUILD_TESTS "Build test executables" OFF)
+            add_subdirectory(third-party/mimalloc EXCLUDE_FROM_ALL)
+        endfunction()
+
+        mold_add_mimalloc()
+    endif()
+
+    add_library(mimalloc::mimalloc ALIAS mimalloc)
+endif()
+
+# Setup TBB
+if(MOLD_USE_SYSTEM_TBB)
+    find_package(TBB REQUIRED)
+else()
+    function(mold_add_tbb)
+        set(CMAKE_BUILD_TYPE RelWithDebInfo)
+        set(BUILD_SHARED_LIBS OFF)
+        option(TBB_TEST "Enable testing" OFF)
+        string(APPEND CMAKE_CXX_FLAGS -D__TBB_DYNAMIC_LOAD_ENABLED=0)
+        option(TBB_STRICT "Treat compiler warnings as errors" OFF)
+        add_subdirectory(third-party/tbb EXCLUDE_FROM_ALL)
+    endfunction()
+
+    mold_add_tbb()
+endif()
+
+if(MOLD_USE_LTO)
+    include(CheckIPOSupported)
+    check_ipo_supported()
+endif()
+
+include(CheckPIESupported)
+check_pie_supported()
+
+add_library(mold-wrapper SHARED)
+set_target_properties(mold-wrapper PROPERTIES
+    POSITION_INDEPENDENT_CODE ON
+    PREFIX "" # Remove the default `lib` prefix
+    INTERPROCEDURAL_OPTIMIZATION ${MOLD_USE_LTO})
+target_link_libraries(mold-wrapper PRIVATE ${CMAKE_DL_LIBS})
+target_sources(mold-wrapper PRIVATE elf/mold-wrapper.c)
+
+add_executable(mold)
+target_compile_features(mold PRIVATE cxx_std_20)
+set_target_properties(mold PROPERTIES
+    CXX_EXTENSIONS OFF
+    INTERPROCEDURAL_OPTIMIZATION ${MOLD_USE_LTO})
+target_include_directories(mold PRIVATE third-party)
+
+if(NOT MSVC)
+    target_compile_options(mold PRIVATE
+        $<$<PLATFORM_ID:Android>:-Wno-c++11-narrowing>
+        -fno-exceptions
+        -fno-unwind-tables
+        -fno-asynchronous-unwind-tables)
+
+    find_package(ZLIB REQUIRED)
+    find_library(M_LIBRARIES m REQUIRED)
+    target_link_libraries(mold PRIVATE
+        ZLIB::ZLIB
+        ${M_LIBRARIES})
+endif()
+
+target_link_libraries(mold PRIVATE
+    TBB::tbb
+    $<$<BOOL:MOLD_USE_MIMALLOC>:mimalloc::mimalloc>)
+
+if(CMAKE_SYSTEM_NAME MATCHES Linux)
+    find_package(OpenSSL REQUIRED COMPONENTS Crypto)
+    target_link_libraries(mold PRIVATE OpenSSL::Crypto)
+endif()
+
+if(CMAKE_SYSTEM_PROCESSOR MATCHES riscv64
+    OR CMAKE_SYSTEM_PROCESSOR MATCHES armv6)
+    find_library(ATOMIC_LIBRARIES atomic REQUIRED)
+    target_link_libraries(mold PRIVATE ${ATOMIC_LIBRARIES})
+endif()
+
+file(STRINGS .git/HEAD GIT_HASH)
+
+if(GIT_HASH MATCHES "ref: ")
+    string(REPLACE "ref: " "" GIT_HASH "${GIT_HASH}")
+    file(STRINGS .git/${GIT_HASH} GIT_HASH)
+endif()
+
+target_compile_definitions(mold PRIVATE
+    GIT_HASH="${GIT_HASH}"
```

Comment:
This will cause nodes in the build graph to be stale whenever the git hash changes. This leads to unnecessary compiles when building incrementally.

@rui314 This hash thing should really be done in a way that makes the binary deterministic and not dependent on git. I played with this locally and you can add a placeholder in the binary like so:

```cpp
char const git_hash_buffer[] = "\x00" "patch git hash here patch git hash here";
auto const git_hash = std::string_view(git_hash_buffer);
```

After that, the resulting binary is reproducible and can be patched before packaging/distribution:

```sh
printf %s "$git_hash" | dd of=mold bs=1 count=40 conv=notrunc "seek=$(grep -aboEe '.patch git hash here patch git hash here' | cut -f1 -d:)"
```

And because the first byte is a null byte, you can even adjust accordingly at runtime.

Reply:
Should this be made a separate issue? I think there is room to improve in this regard.